mlflow/llama_index/tracer.py

```diff
@@ -331,3 +378,95 @@ def _extract_token_usage(
+        return (
+            inspect.isgenerator(result)  # noqa: SIM101
+            or isinstance(result, (StreamingResponse, AsyncStreamingResponse))
+            or isinstance(result, StreamingAgentChatResponse)
+            or (isinstance(result, TaskStepOutput) and self.is_streaming_result(result.output))
+        )
+
+    def register_stream_span(self, span: LiveSpan, result: Any) -> bool:
+        """
+        Register the pending streaming span with the associated generator.
+
+        Args:
+            span: The span that has a streaming output.
+            result: The streaming result that is being processed.
+
+        Returns:
+            True if the span is registered successfully, False otherwise.
+        """
+        if inspect.isgenerator(result):
+            stream = result
+        elif isinstance(result, (StreamingResponse, AsyncStreamingResponse)):
+            stream = result.response_gen
+        elif isinstance(result, StreamingAgentChatResponse):
+            stream = result.chat_stream
+        elif isinstance(result, TaskStepOutput):
+            stream = result.output.chat_stream
+        else:
+            raise ValueError(f"Unsupported streaming response type: {type(result)}")
+
+        if inspect.getgeneratorstate(stream) == inspect.GEN_CLOSED:
+            # Not registering the span because the generator is already exhausted
+            return False
+
+        self._span_id_to_span_and_gen[span.span_id] = (span, stream)
+        return True
+
+    def resolve(self, span: LiveSpan, event: _StreamEndEvent):
+        """
+        Finish the streaming span and recursively resolve the parent spans that
+        returns the same (or derived) stream.
+        """
+        _, stream = self._span_id_to_span_and_gen.pop(span.span_id, (None, None))
+        if not stream:
+            return
+
+        if isinstance(event, (LLMChatEndEvent, LLMCompletionEndEvent)):
+            outputs = event.response
+            status = SpanStatusCode.OK
+        elif isinstance(event, ExceptionEvent):
+            outputs = None
+            status = SpanStatusCode.ERROR
+            span.add_event(SpanEvent.from_exception(event.exception))
+        else:
+            raise ValueError(f"Unsupported event type to resolve streaming: {type(event)}")
+
+        _end_span(client=self._client, span=span, status=status, outputs=outputs)
+
+        # Extract the complete text from the event.
+        if isinstance(outputs, ChatResponse):
+            output_text = outputs.message.content
+        elif isinstance(outputs, CompletionResponse):
+            output_text = outputs.response.text
+        else:
+            output_text = None
+
+        # Recursively resolve the parent spans that are also waiting for the same token
+        # stream to be exhausted.
+        while span.parent_id in self._span_id_to_span_and_gen:
+            if span_and_stream := self._span_id_to_span_and_gen.pop(span.parent_id):
```

Comment:
is it possible for a parent span to create multiple streaming children? if it is, i'm wondering if there would be any problems if one of the streams finished early, and we resolve the parent straight away without waiting for the others to resolve as well.

Reply:
Great point! Yes, a parent span may have multiple child stream spans, but theoretically it works fine, **as long as the second child stream starts after the first child stream completion**. This should be a reasonable assumption in synchronous world, because we cannot yield from two streams at the same time. It does not necessarily true in async stream, which we don't support yet (and not sure we could....).

If the assumption holds, the parent span is only resolved by the second child span, because the parent span is added to `self._span_id_to_span_and_gen` after the first child stream completes.
1. The parent span starts.
2.   The child span 1 starts.
3. The child span 1 returns a stream and adds self to `self._span_id_to_span_and_gen`.
4. The child span 1 finishes consuming the stream. (trying to resolve parent, but at this moment nothing else is in `self._span_id_to_span_and_gen`).
5. The child span 2 starts.
6. Ths child span 2 returns a stream and adds self to `self._span_id_to_span_and_gen`.
7. The parent span returns a stream and adds self to `self._span_id_to_span_and_gen`.
8. The child span 2 finishes consuming the stream. Then resolve the parent span as it's been pended in `self._span_id_to_span_and_gen`.